freeciv/freeciv/doc/man/civserver.6
.\" Freeciv - Copyright (C) 1996 - A Kjeldberg, L Gregersen, P Unold
.\"   This program is free software; you can redistribute it and/or modify
.\"   it under the terms of the GNU General Public License as published by
.\"   the Free Software Foundation; either version 2, or (at your option)
.\"   any later version.
.\"
.\"   This program is distributed in the hope that it will be useful,
.\"   but WITHOUT ANY WARRANTY; without even the implied warranty of
.\"   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\"   GNU General Public License for more details.
.\"
.TH FREECIV 6 "January 16th 2009"
.SH NAME
civserver \- The server for the Freeciv game
.SH SYNOPSIS
.B civserver \
[ \-b|\-\-bind \fIaddress\fP ] \
[ \-d|\-\-debug \fIlevel_number\fP ] \
[ \-e|\-\-exit\-on\-end ] \
[ \-f|\-\-file \fIfilename\fP ] \
[ \-h|\-\-help ] \
[ \-l|\-\-log \fIfilename\fP ] \
[ \-M|\-\-Metaserver \fIaddress\fP ] \
[ \-m|\-\-meta ] \
[ \-p|\-\-port \fIport\fP ] \
[ \-P|\-\-Ppm ] \
[ \-q|\-\-quitidle \fItime\fP ] \
[ \-R|\-\-Ranklog \fIfilename\fP ] \
[ \-r|\-\-read \fIfilename\fP ] \
[ \-S|\-\-Serverid \fIid\fP ] \
[ \-s|\-\-saves \fIdirectory\fP ] \
[ \-v|\-\-version ]

Auth aware servers have additional parameters:
.B [ \-a|\-\-auth ] \
[ \-G|\-\-Guests ] \
[ \-N|\-\-Newusers ]

IPv6 aware servers have additional parameter:
.B [ \-A|\-\-announce \fIprotocol\fP ]

.SH DESCRIPTION
Freeciv is a free turn-based multiplayer strategy game, in which each player
becomes the leader of a civilization, fighting to obtain the ultimate goal:
To become the greatest civilization.

Players of Civilization II\*R by Microprose\*R should feel at home, since one
aim of Freeciv is to have compatible rules. 

Freeciv is maintained by an international team of coders and enthusiasts, and is
easily one of the most fun and addictive network games out there!

This is the server program used to establish a Freeciv server - See also
.IR civclient (6)

This manual page only lists the command line arguments. For details
of the directives necessary to configure
.B civserver
see the Server Manual which can be found at
.UR
http://www.freeciv.org/
.UE
as well as the documentation which is part of the Freeciv distribution.
.SH OPTIONS
The following options are accepted on the command line of the server. They may
not be combined; that is, "civserver \-fp savegame.sav 5557" will not work,
instead you would need to enter "civserver \-f savegame.sav \-p 5557". All
options have a short form (single hyphen and single letter) and a long form
(double hyphen and a complete word); their effects are identical.
.TP
.BI "\-A, \-\-announce \fIprotocol\fP"
Uses \fIprotocol\fP when announcing game in LAN.
Possible values for \fIprotocol\fP are:

\fBIPv4\fP

\fBIPv6\fP

\fBnone\fP

.TP
.BI "\-a, \-\-auth"
Enables server authentication and allows access to registered players only.
.TP
.BI "\-b \fIaddress\fP, \-\-bind \fIaddress\fP"
Sets the \fIaddress\fP on which the server will listen for clients.
.TP
.BI "\-d \fIlevel_number\fP, \-\-debug \fIlevel_number\fP"
Sets the amount of debugging information to be logged in the file named by the
.I \-l
option. Acceptable values for the \fIlevel_number\fP are:

\fB0\fP    for fatal messages only.

\fB1\fP    for fatal and error messages.

\fB2\fP    for fatal, error and normal messages.

\fB3\fP    for fatal, error, normal, and verbose messages.

.TP
.BI "\-e, \-\-exit\-on\-end"
When a game ends, exit instead of restarting.
.TP
.BI "\-f \fIfilename\fP, \-\-file \fIfilename\fP"
Loads a saved game (\fIfilename\fP) into the server before initialization,
instead of starting a new game. This is for reloading saved games, or for 
loading scenarios. The distribution comes with such scenarios, typically stored
in \fI/usr/local/share/freeciv/scenario\fP:

.IP
\(bu
.I british\-isles\-85x80\-v2.80.sav

\(bu
.I earth\-160x90\-v2.sav

\(bu
.I earth\-80x50\-v2.sav

\(bu
.I europe\-200x100\-v2.sav

\(bu
.I hagworld\-120x60\-v1.2.sav

\(bu
.I iberian\-peninsula\-136x100\-v1.0.sav

.TP
.BI "\-G, \-\-Guests"
Allow guests to login as 'guest' when authentication is enabled.
.TP
.TP
.BI "\-h, \-\-help"
Prints out a description of the command line options and exits.
.TP
.BI "\-l \fIfilename\fP, \-\-log \fIfilename\fP"
Defines a log file, \fIfilename\fP, to be produced during processing. Use the
.I \-d
option to set how much is logged.  By default, fatal and normal messages are 
printed to standard output.  With a log file, such messages go to the log
instead.
.TP
.BI "\-M \fIaddress\fP, \-\-Metaserver \fIaddress\fP"
Specifies the \fIaddress\fP of the metaserver to send your server's information
to. 
.TP
.BI "\-m, \-\-meta"
Directs the server to communicate with a metaserver, a place where Freeciv
servers can publish their presence. Freeciv clients consult the metaserver to
offer users a selection of games to join. By default the standard Freeciv 
metaserver is used. You may specify an alternate metaserver with the \fI\-M\fP
option.
.TP
.BI "\-N, \-\-Newusers"
Allow new users to login and be registered in the players base if authentication
is enabled.
.TP
.BI "\-p \fIport\fP, \-\-port \fIport\fP"
Specifies the TCP \fIport\fP number to which clients will connect; players must know
this number to be able to connect if they are not to use the default of 5556
decimal. You may need to use this if 5556 is not available for your use on your
system, or if you would like to run multiple servers on the same system.
.TP
.BI "\-P, \-\-Ppm"
Save map ppms to savegames.
.TP
.BI "\-q \fItime\fP, \-\-quitidle \fItime\fP"
Quits if no players are present for the specified \fItime\fP, in seconds, and 
restarts a new server.
.TP
.BI "\-R \fIfilename\fP, \-\-Ranklog \fIfilename\fP"
Defines a log file, \fIfilename\fP, to be produced during processing. This log
contains information for player ranking.
.TP
.BI "\-r \fIfilename\fP, \-\-read \fIfilename\fP"
Specifies a file, \fIfilename\fP, of server commands which the server will 
automatically read and process, as if you had typed them in at the server's 
prompt. The distribution comes with a few examples that set options to be 
similar to Civilization I\*R and Civilization II\*R . The files are named 
\fIciv1.serv\fP and \fIciv2.serv\fP, and are typically found at 
\fI/usr/local/share/freeciv/\fP.
.TP
.BI "\-S, \-\-Serverid \fIid\fP"
Sets the server \fIid\fP.
.TP
.BI "\-s \fIdirectory\fP, \-\-saves \fIdirectory\fP"
Specifies the \fIdirectory\fP to place save game files created by the server.
This is especially useful when you're running more than one server on the same
system, as it inhibits your servers from clobbering other's save game files.
.TP
.BI "\-v, \-\-version"
Causes the server to display its version number and exit.
.SH EXAMPLES
.TP
.B civserver \-\-file oldgame.sav \-\-port 2244
Starts a server on port \fI2244\fP, loading the save game file 
\fIoldgame.sav\fP.
.TP
.B civserver \-R ranklog \-l logfile \-r script \-f oldgame.sav.gz -p 2244
Starts a server on port \fI2244\fP, loading the save game file 
\fIoldgame.sav.gz\fP. Ranking related events are written to \fIranklog\fP, other 
logging information is written to \fIlogfile\fP. When the server starts, it 
immediately executes the commands contained in \fIscript\fP.
.TP
.B civserver -m -a -q 60 -p 2244 -d 2 -l logfile -r script -s ~/saves
Starts a server on port \fI2244\fP with authentication enabled and communicates 
its existence to the standard Freeciv metaserver. Fatal, error, and normal 
messages are written to \fIlogfile\fP. When the server starts it immediately 
executes the commands contained in \fIscript\fP. Save game files are stored in 
the \fI~/saves\fP directory. When there are no users on the server for \fI60\fP
seconds, the server will restart.
.SH COMMANDS
You may enter commands into the server at any time, either before or during the
running of a game. 

Type "help" or "help help" for starters.
.SH FILES
The Freeciv server requires the following files in the Freeciv data directory,
which is
.I /usr/local/share/freeciv
by default:

\(bu
.I default/buildings.ruleset

\(bu
.I default/cities.ruleset

\(bu
.I default/game.ruleset

\(bu
.I default/governments.ruleset

\(bu
.I default/nations.ruleset

\(bu
.I default/techs.ruleset

\(bu
.I default/terrain.ruleset

\(bu
.I default/units.ruleset

These are the default rule sets used for the game. Alternate sets of rules can
be used by placing them in a separate directory and using the \fBrulesetdir\fP
command to change your server's ruleset directory. 

Type \fBhelp rulesetdir\fP for more information.
.SH ENVIRONMENT
The Freeciv server accepts these environment variables:
.TP
.BI FREECIV_CAPS
A string containing a list of "capabilities" provided by the server. The
compiled-in default should be correct for most purposes, but if you are familiar
with the capability facility in the source you may use it to enforce some
constraints between clients and server.
.TP
.BI FREECIV_COMPRESSION_LEVEL
Sets the compression level for network traffic.
.TP
.BI FREECIV_DATA_ENCODING
Sets the data encoding (used for data files, savegames, and network strings).
.TP
.BI FREECIV_INTERNAL_ENCODING
Sets the internal encoding (used for GUI strings).
.TP
.BI FREECIV_LOCAL_ENCODING
Sets the local encoding (used for terminal output).
.TP 
.BI FREECIV_MULTICAST_GROUP
Sets the multicast group (for the LAN tab).
.TP
.BI FREECIV_PATH
A colon separated list of directories pointing to the
.B freeciv
data directory. By default freeciv looks in the following directories, in order,
for any data files: the current directory; the "data" subdirectory of the
current directory; the subdirectory ".freeciv" in the user's home directory; and
the directory where the files are placed by running "make install".
.TP
.BI HOME
Specifies the user's home directory.
.TP
.BI http_proxy
Set this variable accordingly when using a proxy.
.TP
\fBLANG\fP  or  \fBLANGUAGE\fP
Sets the language and locale on some platforms.
.TP
\fBLC_ALL\fP  or  \fBLC_CTYPE\fP
Similar to LANG (see documentation for your system).
.TP
.BI USER
Specifies the username of the current user.
.SH BUGS
Please report bugs to
.UR
http://bugs.freeciv.org/
.UE
\.

.SH "MORE INFO"
The Freeciv homepage is located at
.UR
http://www.freeciv.org/
.UE
\.

Updates and new info is first posted there.
.SH AUTHORS
The Freeciv Team <freeciv-dev AT freeciv.org>.

This manpage was originally put together by Florian Ernst 
<florian_ernst AT gmx.net> using the Server Manual and the comments in the 
sourcecode. It was updated by Ben Bettin <bwbettin AT gmail.com> to add new
features, integrate information from the website's online documentation, and 
for slight formatting adjustments. Feel free to use it as you wish.
.SH "SEE ALSO"
.IR civclient (6)
and the Server Manual on the Freeciv homepage.


freeciv/server/civserver.c
/**********************************************************************
 Freeciv - Copyright (C) 1996 - A Kjeldberg, L Gregersen, P Unold
   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2, or (at your option)
   any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.
***********************************************************************/

#ifdef HAVE_CONFIG_H
#include <config.h>
#endif

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#ifdef HAVE_SIGNAL_H
#include <signal.h>
#endif

#ifdef GENERATING_MAC  /* mac header(s) */
#include <Controls.h>
#include <Dialogs.h>
#endif

#ifdef WIN32_NATIVE
#include <windows.h>
#endif

/* utility */
#include "fciconv.h"
#include "fcintl.h"
#include "log.h"
#include "shared.h"
#include "support.h"
#include "timing.h"

/* common */
#include "capstr.h"
#include "game.h"
#include "version.h"

/* server */
#include "console.h"
#include "ggzserver.h"
#include "meta.h"
#include "sernet.h"
#include "srv_main.h"

#include "civserver.h"

#ifdef GENERATING_MAC
static void Mac_options(int argc);  /* don't need argv */
#endif

#ifdef HAVE_SIGNAL_H
#  define USE_INTERRUPT_HANDLERS
#endif

#ifdef USE_INTERRUPT_HANDLERS
/**************************************************************************
  This function is called when a SIGINT (ctrl-c) is received.  It will exit
  only if two SIGINTs are received within a second.

  TODO: SIGHUP and SIGTERM should be handled too.  At a minimum we should
  save the game before exiting.
**************************************************************************/
static void sigint_handler(int sig)
{
  static struct timer *timer = NULL;

  if (with_ggz) {
    exit(EXIT_SUCCESS);
  }
  if (timer && read_timer_seconds(timer) <= 1.0) {
    exit(EXIT_SUCCESS);
  } else {
    if (game.info.timeout == -1) {
      freelog(LOG_NORMAL, _("Setting timeout to 0. Autogame will stop.\n"));
      game.info.timeout = 0;
    }
    if (!timer) {
      freelog(LOG_NORMAL, _("You must interrupt Freeciv twice"
                            " within one second to make it exit.\n"));
    }
  }
  timer = renew_timer_start(timer, TIMER_USER, TIMER_ACTIVE);
}
#endif

/**************************************************************************
 Entry point for Freeciv server.  Basically, does two things:
  1. Parses command-line arguments (possibly dialog, on mac).
  2. Calls the main server-loop routine.
**************************************************************************/
int main(int argc, char *argv[])
{
  int inx;
  bool showhelp = FALSE;
  bool showvers = FALSE;
  char *option = NULL;

  /* Load win32 post-crash debugger */
#ifdef WIN32_NATIVE
# ifndef NDEBUG
  if (LoadLibrary("exchndl.dll") == NULL) {
#  ifdef DEBUG
    fprintf(stderr, "exchndl.dll could not be loaded, no crash debugger\n");
#  endif
  }
# endif
#endif

#ifdef USE_INTERRUPT_HANDLERS
  signal(SIGINT, sigint_handler);
#endif

  /* initialize server */
  srv_init();

  /* parse command-line arguments... */

#ifdef GENERATING_MAC
  Mac_options(argc);
#endif
  srvarg.announce = ANNOUNCE_DEFAULT;

  /* no  we don't use GNU's getopt or even the "standard" getopt */
  /* yes we do have reasons ;)                                   */
  inx = 1;
  while (inx < argc) {
    if ((option = get_option_malloc("--file", argv, &inx, argc))) {
      sz_strlcpy(srvarg.load_filename, option);
      free(option);
    } else if (is_option("--help", argv[inx])) {
      showhelp = TRUE;
      break;
    } else if ((option = get_option_malloc("--log", argv, &inx, argc))) {
      srvarg.log_filename = option; /* Never freed. */
    } else if ((option = get_option_malloc("--Ranklog", argv, &inx, argc))) {
      srvarg.ranklog_filename = option; /* Never freed. */
    } else if (is_option("--nometa", argv[inx])) {
      fc_fprintf(stderr, _("Warning: the %s option is obsolete.  "
			   "Use -m to enable the metaserver.\n"), argv[inx]);
      showhelp = TRUE;
    } else if (is_option("--meta", argv[inx])) {
      srvarg.metaserver_no_send = FALSE;
    } else if ((option = get_option_malloc("--Metaserver",
					 argv, &inx, argc))) {
      sz_strlcpy(srvarg.metaserver_addr, option);
      free(option);
      srvarg.metaserver_no_send = FALSE;      /* --Metaserver implies --meta */
    } else if ((option = get_option_malloc("--identity",
					   argv, &inx, argc))) {
      sz_strlcpy(srvarg.metaserver_name, option);
      free(option);
    } else if ((option = get_option_malloc("--port", argv, &inx, argc))) {
      if (sscanf(option, "%d", &srvarg.port) != 1) {
	showhelp = TRUE;
	break;
      }
      free(option);
    } else if ((option = get_option_malloc("--bind", argv, &inx, argc))) {
      srvarg.bind_addr = option; /* Never freed. */
    } else if ((option = get_option_malloc("--topic", argv, &inx, argc))) {
      sz_strlcpy(game.server.meta_info.topic, option);

    } else if ((option = get_option_malloc("--read", argv, &inx, argc)))
      srvarg.script_filename = option; /* Never freed. */
    else if ((option = get_option_malloc("--quitidle", argv, &inx, argc))) {
      if (sscanf(option, "%d", &srvarg.quitidle) != 1) {
	showhelp = TRUE;
	break;
      }
      free(option);
    } else if (is_option("--exit-on-end", argv[inx])) {
      srvarg.exit_on_end = TRUE;
    } else if ((option = get_option_malloc("--debug", argv, &inx, argc))) {
      srvarg.loglevel = log_parse_level_str(option);
      if (srvarg.loglevel == -1) {
	srvarg.loglevel = LOG_NORMAL;
	showhelp = TRUE;
	break;
      }
      free(option);
#ifdef HAVE_AUTH
    } else if ((option = get_option_malloc("--auth", argv, &inx, argc))) {
      srvarg.auth_enabled = TRUE;
      srvarg.auth_conf = option;
    } else if (is_option("--Guests", argv[inx])) {
      srvarg.auth_allow_guests = TRUE;
    } else if (is_option("--Newusers", argv[inx])) {
      srvarg.auth_allow_newusers = TRUE;
#endif
    } else if (is_option("--Ppm", argv[inx])) {
      srvarg.save_ppm = TRUE;
    } else if ((option = get_option_malloc("--Serverid", argv, &inx, argc))) {
      sz_strlcpy(srvarg.serverid, option);
      free(option);
    } else if ((option = get_option_malloc("--saves", argv, &inx, argc))) {
      srvarg.saves_pathname = option; /* Never freed. */
    } else if ((option = get_option_malloc("--scenarios", argv, &inx, argc))) {
      srvarg.scenarios_pathname = option; /* Never freed */
    } else if (is_option("--version", argv[inx])) {
      showvers = TRUE;
    } else if ((option = get_option_malloc("--Announce", argv, &inx, argc))) {
      if (!strcasecmp(option, "ipv4")) {
        srvarg.announce = ANNOUNCE_IPV4;
      } else if(!strcasecmp(option, "none")) {
        srvarg.announce= ANNOUNCE_NONE;
#ifdef IPV6_SUPPORT
      } else if (!strcasecmp(option, "ipv6")) {
        srvarg.announce = ANNOUNCE_IPV6;
#endif /* IPv6 support */
      } else {
        freelog(LOG_ERROR, _("Illegal value \"%s\" for --Announce"), option);
      }
      free(option);
    } else {
      fc_fprintf(stderr, _("Error: unknown option '%s'\n"), argv[inx]);
      showhelp = TRUE;
      break;
    }
    inx++;
  }

  if (showvers && !showhelp) {
    fc_fprintf(stderr, "%s \n", freeciv_name_version());
    exit(EXIT_SUCCESS);
  }
  con_write(C_VERSION, _("This is the server for %s"), freeciv_name_version());
  /* TRANS: No full stop after the URL, could cause confusion. */
  con_write(C_COMMENT, _("You can learn a lot about Freeciv at %s"),
	    WIKI_URL);

  if (showhelp) {
    fc_fprintf(stderr,
	       _("Usage: %s [option ...]\nValid options are:\n"), argv[0]);
    fc_fprintf(stderr, _("  -A  --Announce PROTO\tAnnounce game in LAN using protocol PROTO (IPv4/IPv6/none)\n"));
#ifdef HAVE_AUTH
    fc_fprintf(stderr, _("  -a  --auth FILE\tEnable server authentication "
                         "with configuration from FILE.\n"));
    fc_fprintf(stderr, _("  -G  --Guests\t\tAllow guests to "
			 "login if auth is enabled.\n"));
    fc_fprintf(stderr, _("  -N  --Newusers\tAllow new users to "
			 "login if auth is enabled.\n"));
#endif
    fc_fprintf(stderr, _("  -b  --bind ADDR\tListen for clients on ADDR\n"));
#ifdef DEBUG
    fc_fprintf(stderr, _("  -d, --debug NUM\tSet debug log level (0 to 4,"
		      " or 4:file1,min,max:...)\n"));
#else
    fc_fprintf(stderr,
	       _("  -d, --debug NUM\tSet debug log level (0 to 3)\n"));
#endif
    fc_fprintf(stderr, _("  -f, --file FILE\tLoad saved game FILE\n"));
    fc_fprintf(stderr,
	       _("  -h, --help\t\tPrint a summary of the options\n"));
   fc_fprintf(stderr, _("  -i, --identity ADDR\tBe known as ADDR at metaserver\n"));
    fc_fprintf(stderr, _("  -l, --log FILE\tUse FILE as logfile\n"));
    fc_fprintf(stderr, _("  -m, --meta\t\tNotify metaserver and "
			 "send server's info\n"));
    fc_fprintf(stderr, _("  -M, --Metaserver ADDR\tSet ADDR "
			 "as metaserver address\n"));

    fc_fprintf(stderr, _("  -p, --port PORT\tListen for clients on "
			 "port PORT\n"));
    fc_fprintf(stderr, _("  -q, --quitidle TIME\tQuit if no players "
			 "for TIME seconds\n"));
    fc_fprintf(stderr, _("  -e, --exit-on-end\t"
		      "When a game ends, exit instead of restarting\n"));
    fc_fprintf(stderr,
	       _("  -s, --saves DIR\tSave games to directory DIR\n"));
    fc_fprintf(stderr,
	       _("  -S, --Serverid ID\tSets the server id to ID\n"));
    fc_fprintf(stderr,
	     _("  -P, --Ppm\t\tSave ppms of the map when saving the game.\n"));
    fc_fprintf(stderr, _("  -r, --read FILE\tRead startup script FILE\n"));
    fc_fprintf(stderr,
	       _("  -R, --Ranklog FILE\tUse FILE as ranking logfile\n"));
    fc_fprintf(stderr, _("  -t, --topic \"MESSAGE\"\t\tSet the metaserver topic\n"));

    fc_fprintf(stderr, _("  -v, --version\t\tPrint the version number\n"));
    /* TRANS: No full stop after the URL, could cause confusion. */
    fc_fprintf(stderr, _("Report bugs at %s\n"), BUG_URL);
    exit(EXIT_SUCCESS);
  }

  /* disallow running as root -- too dangerous */
  dont_run_as_root(argv[0], "freeciv_server");

  ggz_initialize();
  init_our_capability();

  /* have arguments, call the main server loop... */
  srv_main();

  /* Technically, we won't ever get here. We exit via server_quit. */

  /* done */
  exit(EXIT_SUCCESS);
}

#ifdef GENERATING_MAC
/*************************************************************************
  generate an option dialog if no options have been passed in
*************************************************************************/
static void Mac_options(int argc)
{
#define HARDCODED_OPT
  /*temporary hack since GetNewDialog() doesn't want to work*/
#ifdef HARDCODED_OPT
  srvarg.log_filename="log.out";
  srvarg.loglevel=LOG_DEBUG;
#else  /* HARDCODED_OPT */
  if (argc == 0)
  {
    OSErr err;
    DialogPtr  optptr;
    Ptr storage;
    Handle ditl;
    Handle dlog;
    short the_type;
    Handle the_handle;
    Rect the_rect;
    short the_item, old_item=16;
    int done = false;
    Str255 the_string;
    /*load/init the stuff for the dialog*/
    storage =NewPtr(sizeof(DialogRecord));
    if (storage == 0)
    {
      exit(EXIT_FAILURE);
    }
    ditl=Get1Resource('DITL',200);
    if ((ditl == 0)||(ResError()))
    {
      exit(EXIT_FAILURE);
    }
    dlog=Get1Resource('DLOG',200);
    if ((dlog == 0)||(ResError()))
    {
      exit(EXIT_FAILURE);
    }
    /*make the dialog*/
    optptr=GetNewDialog(200, storage, (WindowPtr)-1L);
    /*setup the dialog*/
    err=SetDialogDefaultItem(optptr, 1);
    if (err != 0)
    {
      exit(EXIT_FAILURE);
    }
    /*insert default highlight draw code?*/
    err=SetDialogCancelItem(optptr, 2);
    if (err != 0)
    {
      exit(EXIT_FAILURE);
    }
    err=SetDialogTracksCursor(optptr, true);
    if (err != 0)
    {
      exit(EXIT_FAILURE);
    }
    GetDItem(optptr, 16/*normal radio button*/, &the_type, &the_handle, &the_rect);
    SetCtlValue((ControlHandle)the_handle, true);

    while(!done)/*loop*/
    {
      ModalDialog(0L, &the_item);/* don't feed 0 where a upp is expected? */
      	/* old book suggests using OL(NIL) as the first argument, but
           It doesn't include anything about UPPs either, so... */
      switch (the_item)
      {
        case 1:
          done = true;
        break;
        case 2:
          exit(EXIT_SUCCESS);
        break;
        case 13:
          GetDItem(optptr, 13, &the_type, &the_handle, &the_rect);
          srvarg.metaserver_no_send=GetCtlValue((ControlHandle)the_handle);
          SetCtlValue((ControlHandle)the_handle, !srvarg.metaserver_no_send);
        break;
        case 15:
        case 16:
        case 17:
          GetDItem(optptr, old_item, &the_type, &the_handle, &the_rect);
          SetCtlValue((ControlHandle)the_handle, false);
          old_item=the_item;
          GetDItem(optptr, the_item, &the_type, &the_handle, &the_rect);
          SetCtlValue((ControlHandle)the_handle, true);
        break;
      }
    }
    /* now, load the dialog items into the correct variables interpritation */
    GetDItem( optptr, 4, &the_type, &the_handle, &the_rect);
    GetIText( the_handle, (unsigned char *)srvarg.load_filename);
    GetDItem( optptr, 8, &the_type, &the_handle, &the_rect);
    GetIText( the_handle, (unsigned char *)srvarg.log_filename);
    GetDItem( optptr, 12, &the_type, &the_handle, &the_rect);
    GetIText( the_handle, the_string);
    sscanf(the_string, "%d", srvarg.port);
    GetDItem( optptr, 10, &the_type, &the_handle, &the_rect);
    GetIText( the_handle, (unsigned char *)srvarg.script_filename);
    GetDItem(optptr, 15, &the_type, &the_handle, &the_rect);
    if(GetControlValue((ControlHandle)the_handle))
    {
      srvarg.loglevel=LOG_FATAL;
    }
    GetDItem(optptr, 16, &the_type, &the_handle, &the_rect);
    if(GetControlValue((ControlHandle)the_handle))
    {
      srvarg.loglevel=LOG_NORMAL;
    }
    GetDItem(optptr, 17, &the_type, &the_handle, &the_rect);
    if(GetControlValue((ControlHandle)the_handle))
    {
      srvarg.loglevel=LOG_VERBOSE;
    }
    DisposeDialog(optptr);/*get rid of the dialog after sorting out the options*/
    DisposePtr(storage);/*clean up the allocated memory*/
  }
#endif /* HARDCODED_OPT */
#undef  HARDCODED_OPT
}
#endif /* GENERATING_MAC */
